File: code/mathematical_algorithms/src/automorphic_numbers/automorphic_numbers.cpp
#include <iostream>
#include <string>
#include <algorithm> // For std::equal

using namespace std;

/*
 * In mathematics an automorphic number (sometimes referred to as a circular number) is a number
 * whose square "ends" in the same digits as the number itself.
 * For example, 5^2 = 25, 6^2 = 36, 76^2 = 5776, 376^2 = 141376,and 890625^2 = 793212890625,
 * so 5, 6, 76 and 890625
 */

// Part of Cosmos by OpenGenus Foundation
int main()
{
    long long int number; // May have to deal with large numbers
    cout << "Enter a number: ";
    cin >> number;

    unsigned long long int numberSquared = number * number; // Squares can get really big

    string numberString = to_string(number);
    string numberSquaredString = to_string(numberSquared);

    bool isAutomorphic = std::equal(numberString.rbegin(),
                                    numberString.rend(), numberSquaredString.rbegin());                        // Checks if the number is at the end of the number squared

    cout << number << " is " << (isAutomorphic ? "" : "not ") << "an automorphic number." << endl;
}

File: code/mathematical_algorithms/src/automorphic_numbers/automorphic_numbers.cs
/* Part of Cosmos by OpenGenus Foundation */
using System;

namespace AutomophicCSharp
{
    class Automorphic
    {
        static void Main (string[] args)
        {
            for(double i=0; i<=10000000; i++)
            {
                if(isAutomorphic(i))
                {
                    Console.WriteLine(String.Format("{0} is automorphic - {1}", i, i*i));
                }
            }
        }

        static bool isAutomorphic(double value)
        {
            // convert the value to a string
            string stringValue = value.ToString();

            // get the square as a string
            string stringSquared = (value * value).ToString();

            return stringSquared.EndsWith(stringValue);
        }
    }
}